Base/usr/share/man/man2/futex.md

```diff
@@ -0,0 +1,191 @@
+simple "event" synchronization primitive. An event has a boolean state: it can
+be *set* or *unset*; the initial state being unset. The two operations on an
+event are *waiting* until it is set, and *setting* it (which wakes up any
+threads that were waiting for the event to get set).
+
+Such a synchronization primitive could be used, for example, to notify threads
+that are waiting for another thread to perform some sort of complex
+initialization.
+
+The implementation features two fast paths: both setting an event that no
+thread is waiting on, and trying to wait on an event that has already been set,
+are performed entirely in userspace without calling into the kernel. For this
+to work, the value of the futex integer is used to track both the state of the
+event (whether it has been set) and whether any threads are waiting on it.
+
+```c++
+#include <AK/Atomic.h>
+#include <serenity.h>
+
+class Event {
+private:
+    enum State : u32 {
+        UnsetNoWaiters,
+        UnsetWithWaiters,
+        Set,
+    };
+
+    AK::Atomic<State> m_state { UnsetNoWaiters };
+
+public:
+    void set()
+    {
+        State previous_state = m_state.exchange(Set, AK::memory_order_release);
+        // If there was anyone waiting, wake them all up.
+        // Fast path: no one was waiting, so we're done.
+        if (previous_state == UnsetWithWaiters)
+            futex_wake((u32*)m_state.ptr(), UINT32_MAX, false);
+    }
+
+    void wait()
+    {
+        // If the state is UnsetNoWaiters, set it to UnsetWithWaiters.
+        State expected_state = UnsetNoWaiters;
+        bool have_exchanged = m_state.compare_exchange_strong(
+            expected_state, UnsetWithWaiters,
+            AK::memory_order_acquire);
+        if (have_exchanged)
+            expected_state = UnsetWithWaiters;
+
+        // We need to check the state in a loop and not just once
+        // because of the possibility of spurious wakeups.
+        // Fast path: if the state was already Set, we're done.
+        while (expected_state != Set) {
+            futex_wait((u32*)m_state.ptr(), expected_state, nullptr, 0, false);
+            expected_state = m_state.load(AK::memory_order_acquire);
+        }
+    }
+};
+```
+
+## History
+
+The name "futex" stands for "fast userspace mutex".
+
+The `futex()` system call originally apeared in Linux. Since then, many other
+kernels implemented support for futex-like operations, under various names, in
+particular:
+* Darwin (XNU) has private `ulock_wait()` and `ulock_wake()` API;
+* Windows (NT) apparently has `WaitOnAddress()`, `WakeByAddressSingle()` and
+  `WakeByAddressAll()`;
+* FreeBSD and DargonFly BSD have `umtx`;
+* OpenBSD has Linux-like `futex()`;
+* GNU Hurd has `gsync_wait()`, `gsync_wake()`, and `gsync_requeue()`.
+
+## See also
+
+* [Futexes Are Tricky](https://akkadia.org/drepper/futex.pdf) by Ulrich Drepper
+* [Locking in WebKit](https://webkit.org/blog/6161/locking-in-webkit/) by Filip Pizlo
```

Comment:
I'm not sure, but I think See also is reserved for other manpages (and in this case we don't have any related ones, we need manpages for pthread condvars, pthread mutexes, etc.). This can go in something like "Further Reading"